<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>preprocessors - Galaxy-ML Documentation</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "preprocessors";
    var mkdocs_page_input_path = "APIs/preprocessors.md";
    var mkdocs_page_url = "/Galaxy-ML/APIs/preprocessors/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-141380533-1', 'goeckslab.github.io/Galaxy-ML/');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Galaxy-ML Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">APIs</span>
    <ul class="subnav">
                <li class=" current">
                    
    <a class="current" href="./">preprocessors</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#z_randomoversampler">Z_RandomOverSampler</a></li>
    

    <li class="toctree-l3"><a href="#tdmscaler">TDMScaler</a></li>
    

    <li class="toctree-l3"><a href="#genomeonehotencoder">GenomeOneHotEncoder</a></li>
    

    <li class="toctree-l3"><a href="#proteinonehotencoder">ProteinOneHotEncoder</a></li>
    

    <li class="toctree-l3"><a href="#imagebatchgenerator">ImageBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#fastaiterator">FastaIterator</a></li>
    

    <li class="toctree-l3"><a href="#fastatoarrayiterator">FastaToArrayIterator</a></li>
    

    <li class="toctree-l3"><a href="#fastadnabatchgenerator">FastaDNABatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#fastarnabatchgenerator">FastaRNABatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#fastaproteinbatchgenerator">FastaProteinBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#intervalstoarrayiterator">IntervalsToArrayIterator</a></li>
    

    <li class="toctree-l3"><a href="#genomicintervalbatchgenerator">GenomicIntervalBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#genomicvariantbatchgenerator">GenomicVariantBatchGenerator</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../feature-selectors/">feature_selectors</a>
                </li>
                <li class="">
                    
    <a class="" href="../iraps-classifier/">iraps_classifier</a>
                </li>
                <li class="">
                    
    <a class="" href="../binarize-target/">binarize estimators/scorers</a>
                </li>
                <li class="">
                    
    <a class="" href="../keras-galaxy-models/">keras_galaxy_models</a>
                </li>
                <li class="">
                    
    <a class="" href="../model-validations/">model_validations</a>
                </li>
                <li class="">
                    
    <a class="" href="../model-persistent/">model_persistent</a>
                </li>
                <li class="">
                    
    <a class="" href="../utils/">utils</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Releases</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../CHANGELOG/">Change Log</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Galaxy-ML Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>APIs &raquo;</li>
        
      
    
    <li>preprocessors</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/goeckslab/Galaxy-ML/edit/master/docs/APIs/preprocessors.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L30">[source]</a></span></p>
<h2 id="z_randomoversampler">Z_RandomOverSampler</h2>
<pre><code class="python">galaxy_ml.preprocessors.Z_RandomOverSampler(sampling_strategy='auto', return_indices=False, random_state=None, ratio=None, negative_thres=0, positive_thres=-1)
</code></pre>

<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L112">[source]</a></span></p>
<h2 id="tdmscaler">TDMScaler</h2>
<pre><code class="python">galaxy_ml.preprocessors.TDMScaler(q_lower=25.0, q_upper=75.0)
</code></pre>

<p>Scale features using Training Distribution Matching (TDM) algorithm</p>
<p><strong>References</strong></p>
<pre><code>.. [1] Thompson JA, Tan J and Greene CS (2016) Cross-platform
       normalization of microarray and RNA-seq data for machine
       learning applications. PeerJ 4, e1621.
</code></pre>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L195">[source]</a></span></p>
<h2 id="genomeonehotencoder">GenomeOneHotEncoder</h2>
<pre><code class="python">galaxy_ml.preprocessors.GenomeOneHotEncoder(fasta_path=None, padding=True, seq_length=None)
</code></pre>

<p>Convert Genomic sequences to one-hot encoded 2d array</p>
<p><strong>Paramaters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str, default None<br>
        File path to the fasta file. There could two other ways to set up
        <code>fasta_path</code>. 1) through fit_params; 2) set_params(). If fasta_path is
        None, we suppose the sequences are contained in first column of X.</li>
<li><strong>padding</strong>: bool, default is False<br>
        All sequences are expected to be in the same length, but sometimes not.
        If True, all sequences use the same length of first entry by either
        padding or truncating. If False, raise ValueError if different seuqnce
        lengths are found.</li>
<li><strong>seq_length</strong>: None or int<br>
        Sequence length. If None, determined by the the first entry.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L312">[source]</a></span></p>
<h2 id="proteinonehotencoder">ProteinOneHotEncoder</h2>
<pre><code class="python">galaxy_ml.preprocessors.ProteinOneHotEncoder(fasta_path=None, padding=True, seq_length=None)
</code></pre>

<p>Convert protein sequences to one-hot encoded 2d array</p>
<p><strong>Paramaters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str, default None<br>
         File path to the fasta file. There could two other ways to set up
        <code>fasta_path</code>. 1) through fit_params; 2) set_params(). If fasta_path is
        None, we suppose the sequences are contained in first column of X.</li>
<li><strong>padding</strong>: bool, default is False<br>
        All sequences are expected to be in the same length, but sometimes not.
        If True, all sequences use the same length of first entry by either
        padding or truncating. If False, raise ValueError if different seuqnce
        lengths are found.</li>
<li><strong>seq_length</strong>: None or int<br>
        Sequence length. If None, determined by the the first entry.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L380">[source]</a></span></p>
<h2 id="imagebatchgenerator">ImageBatchGenerator</h2>
<pre><code class="python">keras.preprocessing.image.ImageBatchGenerator(featurewise_center=False, samplewise_center=False, featurewise_std_normalization=False, samplewise_std_normalization=False, zca_whitening=False, zca_epsilon=1e-06, rotation_range=0, width_shift_range=0.0, height_shift_range=0.0, brightness_range=None, shear_range=0.0, zoom_range=0.0, channel_shift_range=0.0, fill_mode='nearest', cval=0.0, horizontal_flip=False, vertical_flip=False, rescale=None, preprocessing_function=None, data_format=None, validation_split=0.0, dtype=None)
</code></pre>

<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L390">[source]</a></span></p>
<h2 id="fastaiterator">FastaIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors.FastaIterator(n, batch_size=32, shuffle=True, seed=0)
</code></pre>

<p>Base class for fasta sequence iterators.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>n</strong>: int<br>
        Total number of samples</li>
<li><strong>batch_size</strong>: int<br>
        Size of batch</li>
<li><strong>shuffle</strong>: bool<br>
        Whether to shuffle data between epoch</li>
<li><strong>seed</strong>: int<br>
        Random seed number for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L410">[source]</a></span></p>
<h2 id="fastatoarrayiterator">FastaToArrayIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors.FastaToArrayIterator(X, generator, y=None, batch_size=32, shuffle=True, sample_weight=None, seed=None)
</code></pre>

<p>Iterator yielding Numpy array from fasta sequences</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>X</strong>: array<br>
        Contains sequence indexes in the fasta file</li>
<li><strong>generator</strong>: fitted object<br>
        instance of BatchGenerator, e.g.,  FastaDNABatchGenerator
        or FastaProteinBatchGenerator</li>
<li><strong>y</strong>: array<br>
        Target labels or values</li>
<li><strong>batch_size</strong>: int, default=32<br></li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>sample_weight</strong>: None or array<br>
        Sample weight</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L464">[source]</a></span></p>
<h2 id="fastadnabatchgenerator">FastaDNABatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors.FastaDNABatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L572">[source]</a></span></p>
<h2 id="fastarnabatchgenerator">FastaRNABatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors.FastaRNABatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L600">[source]</a></span></p>
<h2 id="fastaproteinbatchgenerator">FastaProteinBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors.FastaProteinBatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L629">[source]</a></span></p>
<h2 id="intervalstoarrayiterator">IntervalsToArrayIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors.IntervalsToArrayIterator(X, generator, y=None, batch_size=32, shuffle=True, sample_weight=None, seed=None, sample_probabilities=None)
</code></pre>

<p>Iterator yielding Numpy array from intervals and reference
sequences.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>X</strong>: array<br>
        Contains sequence indexes in the fasta file</li>
<li><strong>generator</strong>: fitted object<br>
        instance of GenomicIntervalBatchGenerator.</li>
<li><strong>y</strong>: None<br>
        The existing of y is due to inheritence, should be always None.</li>
<li><strong>batch_size</strong>: int, default=32<br></li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>sample_weight</strong>: None or array<br>
        Sample weight</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
<li><strong>sample_probabilities</strong>: 1-D array or None, default is None.<br>
        The probabilities to draw samples. Different from the sample
        weight, this parameter only changes the the frequency of
        sampling, won't the loss during training.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L699">[source]</a></span></p>
<h2 id="genomicintervalbatchgenerator">GenomicIntervalBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors.GenomicIntervalBatchGenerator(ref_genome_path=None, intervals_path=None, target_path=None, features='infer', blacklist_regions='hg38', shuffle=True, seed=None, seq_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, random_state=None)
</code></pre>

<p>Generate sequence array and target values from a reference
genome, intervals and genomic feature dataset.
Try to mimic the the
<code>selene_sdk.samplers.interval_sampler.IntervalsSampler</code>.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>ref_genome_path</strong>: str<br>
        File path to the reference genomce, usually in fasta
        format.</li>
<li><strong>intervals_path</strong>: str<br>
        File path to the intervals dataset.</li>
<li><strong>target_path</strong>: str<br>
        File path to the dataset containing genomic features or
        target information, usually in <code>bed</code> ir <code>bed.gz</code> format.</li>
<li><strong>features</strong>: list of str or 'infer'<br>
        A list of features to predict. If 'infer', retrieve all
        the unique features from the target file.</li>
<li><strong>blacklist_regions</strong>: str<br>
        E.g., 'hg38'. For more info, refer to
        <code>selene_sdk.sequences.Genome</code>.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs.</li>
<li><strong>seed</strong>: int or None, default=None<br>
        Random seed for shuffling between epocks.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Retrived sequence length.</li>
<li><strong>center_bin_to_predict</strong>: int, default=200<br>
        Query the tabix-indexed file for a region of length
        <code>center_bin_to_predict</code>.</li>
<li><strong>feature_thresholds</strong>: float, default=0.5<br>
        Threshold values to determine target value.</li>
<li><strong>random_state</strong>: int or None, default=None<br>
        Random seed for sampling sequences with changing position.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors.py#L951">[source]</a></span></p>
<h2 id="genomicvariantbatchgenerator">GenomicVariantBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors.GenomicVariantBatchGenerator(ref_genome_path=None, vcf_path=None, blacklist_regions='hg38', seq_length=1000, output_reference=False)
</code></pre>

<p><code>keras.utils.Sequence</code> capable sequence array generator
from a reference genome and VCF (variant call format) file.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>ref_genome_path</strong>: str<br>
        File path to the reference genomce, usually in fasta
        format.</li>
<li><strong>vcf_path</strong>: str<br>
        File path to the VCF dataset.</li>
<li><strong>blacklist_regions</strong>: str<br>
        E.g., 'hg38'. For more info, refer to
        <code>selene_sdk.sequences.Genome</code>.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Retrived sequence length.</li>
<li><strong>output_reference</strong>: bool, default is False.<br>
        If True, output reference sequence instead.</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../feature-selectors/" class="btn btn-neutral float-right" title="feature_selectors">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../.." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/goeckslab/Galaxy-ML/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../.." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../feature-selectors/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
