<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../../img/favicon.ico">
  <title>preprocessors - Galaxy-ML Documentation</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "preprocessors";
    var mkdocs_page_input_path = "APIs/preprocessors.md";
    var mkdocs_page_url = "/Galaxy-ML/APIs/preprocessors/";
  </script>
  
  <script src="../../js/jquery-2.1.1.min.js" defer></script>
  <script src="../../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
  <script>
      (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
      (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
      m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
      })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

      ga('create', 'UA-141380533-1', 'goeckslab.github.io/Galaxy-ML/');
      ga('send', 'pageview');
  </script>
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href="../.." class="icon icon-home"> Galaxy-ML Documentation</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="../..">Home</a>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">APIs</span>
    <ul class="subnav">
                <li class=" current">
                    
    <a class="current" href="./">preprocessors</a>
    <ul class="subnav">
            
    <li class="toctree-l3"><a href="#z_randomoversampler">Z_RandomOverSampler</a></li>
    

    <li class="toctree-l3"><a href="#tdmscaler">TDMScaler</a></li>
    

    <li class="toctree-l3"><a href="#genomeonehotencoder">GenomeOneHotEncoder</a></li>
    

    <li class="toctree-l3"><a href="#proteinonehotencoder">ProteinOneHotEncoder</a></li>
    

    <li class="toctree-l3"><a href="#fastaiterator">FastaIterator</a></li>
    

    <li class="toctree-l3"><a href="#fastatoarrayiterator">FastaToArrayIterator</a></li>
    

    <li class="toctree-l3"><a href="#fastadnabatchgenerator">FastaDNABatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#fastarnabatchgenerator">FastaRNABatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#fastaproteinbatchgenerator">FastaProteinBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#intervalstoarrayiterator">IntervalsToArrayIterator</a></li>
    

    <li class="toctree-l3"><a href="#genomicintervalbatchgenerator">GenomicIntervalBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#genomicvariantbatchgenerator">GenomicVariantBatchGenerator</a></li>
    

    <li class="toctree-l3"><a href="#imagedataframebatchgenerator">ImageDataFrameBatchGenerator</a></li>
    

    </ul>
                </li>
                <li class="">
                    
    <a class="" href="../feature-selectors/">feature_selectors</a>
                </li>
                <li class="">
                    
    <a class="" href="../iraps-classifier/">iraps_classifier</a>
                </li>
                <li class="">
                    
    <a class="" href="../binarize-target/">binarize estimators/scorers</a>
                </li>
                <li class="">
                    
    <a class="" href="../keras-galaxy-models/">keras_galaxy_models</a>
                </li>
                <li class="">
                    
    <a class="" href="../model-validations/">model_validations</a>
                </li>
                <li class="">
                    
    <a class="" href="../model-persistent/">model_persistent</a>
                </li>
                <li class="">
                    
    <a class="" href="../utils/">utils</a>
                </li>
    </ul>
	    </li>
          
            <li class="toctree-l1">
		
    <span class="caption-text">Releases</span>
    <ul class="subnav">
                <li class="">
                    
    <a class="" href="../../CHANGELOG/">Change Log</a>
                </li>
    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="../..">Galaxy-ML Documentation</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="../..">Docs</a> &raquo;</li>
    
      
        
          <li>APIs &raquo;</li>
        
      
    
    <li>preprocessors</li>
    <li class="wy-breadcrumbs-aside">
      
        <a href="https://github.com/goeckslab/Galaxy-ML/edit/master/docs/APIs/preprocessors.md"
          class="icon icon-github"> Edit on GitHub</a>
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_z_random_over_sampler.py#L12">[source]</a></span></p>
<h2 id="z_randomoversampler">Z_RandomOverSampler</h2>
<pre><code class="python">galaxy_ml.preprocessors._z_random_over_sampler.Z_RandomOverSampler(sampling_strategy='auto', return_indices=False, random_state=None, ratio=None, negative_thres=0, positive_thres=-1)
</code></pre>

<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_tdm_scaler.py#L9">[source]</a></span></p>
<h2 id="tdmscaler">TDMScaler</h2>
<pre><code class="python">galaxy_ml.preprocessors._tdm_scaler.TDMScaler(q_lower=25.0, q_upper=75.0)
</code></pre>

<p>Scale features using Training Distribution Matching (TDM) algorithm</p>
<p><strong>References</strong></p>
<pre><code>.. [1] Thompson JA, Tan J and Greene CS (2016) Cross-platform
       normalization of microarray and RNA-seq data for machine
       learning applications. PeerJ 4, e1621.
</code></pre>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_genome_one_hot_encoder.py#L8">[source]</a></span></p>
<h2 id="genomeonehotencoder">GenomeOneHotEncoder</h2>
<pre><code class="python">galaxy_ml.preprocessors._genome_one_hot_encoder.GenomeOneHotEncoder(fasta_path=None, padding=True, seq_length=None)
</code></pre>

<p>Convert Genomic sequences to one-hot encoded 2d array</p>
<p><strong>Paramaters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str, default None<br>
        File path to the fasta file. There could two other ways to set up
        <code>fasta_path</code>. 1) through fit_params; 2) set_params(). If fasta_path is
        None, we suppose the sequences are contained in first column of X.</li>
<li><strong>padding</strong>: bool, default is False<br>
        All sequences are expected to be in the same length, but sometimes not.
        If True, all sequences use the same length of first entry by either
        padding or truncating. If False, raise ValueError if different seuqnce
        lengths are found.</li>
<li><strong>seq_length</strong>: None or int<br>
        Sequence length. If None, determined by the the first entry.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_protein_one_hot_encoder.py#L7">[source]</a></span></p>
<h2 id="proteinonehotencoder">ProteinOneHotEncoder</h2>
<pre><code class="python">galaxy_ml.preprocessors._genome_one_hot_encoder.ProteinOneHotEncoder(fasta_path=None, padding=True, seq_length=None)
</code></pre>

<p>Convert protein sequences to one-hot encoded 2d array</p>
<p><strong>Paramaters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str, default None<br>
         File path to the fasta file. There could two other ways to set up
        <code>fasta_path</code>. 1) through fit_params; 2) set_params(). If fasta_path is
        None, we suppose the sequences are contained in first column of X.</li>
<li><strong>padding</strong>: bool, default is False<br>
        All sequences are expected to be in the same length, but sometimes not.
        If True, all sequences use the same length of first entry by either
        padding or truncating. If False, raise ValueError if different seuqnce
        lengths are found.</li>
<li><strong>seq_length</strong>: None or int<br>
        Sequence length. If None, determined by the the first entry.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_fasta_iterator.py#L9">[source]</a></span></p>
<h2 id="fastaiterator">FastaIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors._fasta_iterator.FastaIterator(n, batch_size=32, shuffle=True, seed=0)
</code></pre>

<p>Base class for fasta sequence iterators.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>n</strong>: int<br>
        Total number of samples</li>
<li><strong>batch_size</strong>: int<br>
        Size of batch</li>
<li><strong>shuffle</strong>: bool<br>
        Whether to shuffle data between epoch</li>
<li><strong>seed</strong>: int<br>
        Random seed number for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_fasta_iterator.py#L29">[source]</a></span></p>
<h2 id="fastatoarrayiterator">FastaToArrayIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors._fasta_iterator.FastaToArrayIterator(X, generator, y=None, batch_size=32, shuffle=True, sample_weight=None, seed=None)
</code></pre>

<p>Iterator yielding Numpy array from fasta sequences</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>X</strong>: array<br>
        Contains sequence indexes in the fasta file</li>
<li><strong>generator</strong>: fitted object<br>
        instance of BatchGenerator, e.g.,  FastaDNABatchGenerator
        or FastaProteinBatchGenerator</li>
<li><strong>y</strong>: array<br>
        Target labels or values</li>
<li><strong>batch_size</strong>: int, default=32<br></li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>sample_weight</strong>: None or array<br>
        Sample weight</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_fasta_dna_batch_generator.py#L9">[source]</a></span></p>
<h2 id="fastadnabatchgenerator">FastaDNABatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._fasta_dna_batch_generator.FastaDNABatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_fasta_rna_batch_generator.py#L4">[source]</a></span></p>
<h2 id="fastarnabatchgenerator">FastaRNABatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._fasta_rna_batch_generator.FastaRNABatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_fasta_protein_batch_generator.py#L4">[source]</a></span></p>
<h2 id="fastaproteinbatchgenerator">FastaProteinBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._fasta_protein_batch_generator.FastaProteinBatchGenerator(fasta_path, seq_length=1000, shuffle=True, seed=None)
</code></pre>

<p>Fasta squence batch data generator, online transformation of sequences
to array.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>fasta_path</strong>: str<br>
        File path to fasta file.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Sequence length, number of bases.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_genomic_interval_batch_generator.py#L9">[source]</a></span></p>
<h2 id="intervalstoarrayiterator">IntervalsToArrayIterator</h2>
<pre><code class="python">galaxy_ml.preprocessors._genomic_interval_batch_generator.IntervalsToArrayIterator(X, generator, y=None, batch_size=32, shuffle=True, sample_weight=None, seed=None, sample_probabilities=None)
</code></pre>

<p>Iterator yielding Numpy array from intervals and reference
sequences.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>X</strong>: array<br>
        Contains sequence indexes in the fasta file</li>
<li><strong>generator</strong>: fitted object<br>
        instance of GenomicIntervalBatchGenerator.</li>
<li><strong>y</strong>: None<br>
        The existing of y is due to inheritence, should be always None.</li>
<li><strong>batch_size</strong>: int, default=32<br></li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs</li>
<li><strong>sample_weight</strong>: None or array<br>
        Sample weight</li>
<li><strong>seed</strong>: int<br>
        Random seed for data shuffling</li>
<li><strong>sample_probabilities</strong>: 1-D array or None, default is None.<br>
        The probabilities to draw samples. Different from the sample
        weight, this parameter only changes the the frequency of
        sampling, won't the loss during training.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_genomic_interval_batch_generator.py#L79">[source]</a></span></p>
<h2 id="genomicintervalbatchgenerator">GenomicIntervalBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._genomic_interval_batch_generator.GenomicIntervalBatchGenerator(ref_genome_path=None, intervals_path=None, target_path=None, features='infer', blacklist_regions='hg38', shuffle=True, seed=None, seq_length=1000, center_bin_to_predict=200, feature_thresholds=0.5, random_state=None)
</code></pre>

<p>Generate sequence array and target values from a reference
genome, intervals and genomic feature dataset.
Try to mimic the the
<code>selene_sdk.samplers.interval_sampler.IntervalsSampler</code>.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>ref_genome_path</strong>: str<br>
        File path to the reference genomce, usually in fasta
        format.</li>
<li><strong>intervals_path</strong>: str<br>
        File path to the intervals dataset.</li>
<li><strong>target_path</strong>: str<br>
        File path to the dataset containing genomic features or
        target information, usually in <code>bed</code> ir <code>bed.gz</code> format.</li>
<li><strong>features</strong>: list of str or 'infer'<br>
        A list of features to predict. If 'infer', retrieve all
        the unique features from the target file.</li>
<li><strong>blacklist_regions</strong>: str<br>
        E.g., 'hg38'. For more info, refer to
        <code>selene_sdk.sequences.Genome</code>.</li>
<li><strong>shuffle</strong>: bool, default=True<br>
        Whether to shuffle the data between epochs.</li>
<li><strong>seed</strong>: int or None, default=None<br>
        Random seed for shuffling between epocks.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Retrived sequence length.</li>
<li><strong>center_bin_to_predict</strong>: int, default=200<br>
        Query the tabix-indexed file for a region of length
        <code>center_bin_to_predict</code>.</li>
<li><strong>feature_thresholds</strong>: float, default=0.5<br>
        Threshold values to determine target value.</li>
<li><strong>random_state</strong>: int or None, default=None<br>
        Random seed for sampling sequences with changing position.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_genomic_variant_batch_generator.py#L10">[source]</a></span></p>
<h2 id="genomicvariantbatchgenerator">GenomicVariantBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._genomic_variant_batch_generator.GenomicVariantBatchGenerator(ref_genome_path=None, vcf_path=None, blacklist_regions='hg38', seq_length=1000, output_reference=False)
</code></pre>

<p><code>keras.utils.Sequence</code> capable sequence array generator
from a reference genome and VCF (variant call format) file.</p>
<p><strong>Parameters</strong></p>
<ul>
<li><strong>ref_genome_path</strong>: str<br>
        File path to the reference genomce, usually in fasta
        format.</li>
<li><strong>vcf_path</strong>: str<br>
        File path to the VCF dataset.</li>
<li><strong>blacklist_regions</strong>: str<br>
        E.g., 'hg38'. For more info, refer to
        <code>selene_sdk.sequences.Genome</code>.</li>
<li><strong>seq_length</strong>: int, default=1000<br>
        Retrived sequence length.</li>
<li><strong>output_reference</strong>: bool, default is False.<br>
        If True, output reference sequence instead.</li>
</ul>
<hr />
<p><span style="float:right;"><a href="https://github.com/goeckslab/Galaxy-ML/blob/master/galaxy_ml/preprocessors/_image_batch_generator.py#L144">[source]</a></span></p>
<h2 id="imagedataframebatchgenerator">ImageDataFrameBatchGenerator</h2>
<pre><code class="python">galaxy_ml.preprocessors._image_batch_generator.ImageDataFrameBatchGenerator(dataframe, featurewise_center=False, samplewise_center=False, featurewise_std_normalization=False, samplewise_std_normalization=False, zca_whitening=False, zca_epsilon=1e-06, rotation_range=0, width_shift_range=0.0, height_shift_range=0.0, brightness_range=None, shear_range=0.0, zoom_range=0.0, channel_shift_range=0.0, fill_mode='nearest', cval=0.0, horizontal_flip=False, vertical_flip=False, rescale=None, preprocessing_function=None, data_format='channels_last', interpolation_order=1, dtype='float32', directory=None, x_col='filename', y_col='class', weight_col=None, target_size=(256, 256), color_mode='rgb', classes=None, class_mode='categorical', shuffle=True, seed=None, save_to_dir=None, save_prefix='', save_format='png', interpolation='nearest', fit_sample_size=None)
</code></pre>

<p>Extend <code>keras_preprocessing.image.ImageDataGenerator</code> to work with
DataFrame exclusively, generating batches of tensor data from
images with online augumentation.</p>
<p><strong>Parameters</strong></p>
<pre><code>From `keras_preprocessing.image.ImageDataGenerator`.
</code></pre>
<ul>
<li><strong>featurewise_center</strong>: Boolean.<br>
        Set input mean to 0 over the dataset, feature-wise.</li>
<li><strong>samplewise_center</strong>: Boolean. Set each sample mean to 0.<br></li>
<li><strong>featurewise_std_normalization</strong>: Boolean.<br>
        Divide inputs by std of the dataset, feature-wise.</li>
<li><strong>samplewise_std_normalization</strong>: Boolean. Divide each input by its std.<br></li>
<li><strong>zca_whitening</strong>: Boolean. Apply ZCA whitening.<br></li>
<li><strong>zca_epsilon</strong>: epsilon for ZCA whitening. Default is 1e-6.<br></li>
<li><strong>rotation_range</strong>: Int. Degree range for random rotations.<br></li>
<li><strong>width_shift_range</strong>: Float, 1-D array-like or int.<br></li>
<li><strong>height_shift_range</strong>: Float, 1-D array-like or int.<br></li>
<li><strong>brightness_range</strong>: Tuple or list of two floats.<br></li>
<li><strong>shear_range</strong>: Float. Shear Intensity.<br></li>
<li><strong>zoom_range</strong>: Float or [lower, upper].<br></li>
<li><strong>channel_shift_range</strong>: Float. Range for random channel shifts.<br>
    fill_mode: One of {"constant", "nearest", "reflect" or "wrap"}.
        Default is 'nearest'.
        Points outside the boundaries of the input are filled
        according to the given mode:
        - 'constant': kkkkkkkk|abcd|kkkkkkkk (cval=k)
        - 'nearest':  aaaaaaaa|abcd|dddddddd
        - 'reflect':  abcddcba|abcd|dcbaabcd
        - 'wrap':  abcdabcd|abcd|abcdabcd</li>
<li><strong>cval</strong>: Float or Int.<br></li>
<li><strong>horizontal_flip</strong>: Boolean.<br>
        Randomly flip inputs horizontally.</li>
<li><strong>vertical_flip</strong>: Boolean.<br></li>
<li><strong>rescale</strong>: rescaling factor. Defaults to None.<br></li>
<li><strong>preprocessing_function</strong>: function that will be applied on each input.<br>
        The function will run after the image is resized and augmented.
        The function should take one argument:
        one image (Numpy tensor with rank 3),
        and should output a Numpy tensor with the same shape.</li>
<li><strong>data_format</strong>: Image data format,<br>
        either "channels_first" or "channels_last".
        "channels_last" mode means that the images should have shape
        <code>(samples, height, width, channels)</code>,
        "channels_first" mode means that the images should have shape
        <code>(samples, channels, height, width)</code>.
        It defaults to the <code>image_data_format</code> value found in your
        Keras config file at <code>~/.keras/keras.json</code>.
        If you never set it, then it will be "channels_last".</li>
<li><strong>interpolation_order</strong>: Int.<br></li>
<li><strong>dtype</strong>: Dtype to use for the generated arrays. Default is 'float32'.<br></li>
<li><strong>dataframe</strong>: Pandas dataframe containing the filepaths relative to<br>
        <code>directory</code>. From <code>keras_preprocessing.image.ImageDataGenerator.
        flow_from_dataframe</code>.</li>
<li><strong>directory</strong>: string, path to the directory to read images from. If <code>None</code>,<br>
        data in <code>x_col</code> column should be absolute paths.</li>
<li><strong>x_col</strong>: string, column in <code>dataframe</code> that contains the filenames (or<br>
                absolute paths if <code>directory</code> is <code>None</code>).</li>
<li><strong>y_col</strong>: string or list, column/s in <code>dataframe</code> that has the target data.<br></li>
<li><strong>weight_col</strong>: string, column in <code>dataframe</code> that contains the sample<br>
        weights. Default: <code>None</code>.</li>
<li><strong>target_size</strong>: tuple of integers <code>(height, width)</code>, default: <code>(256, 256)</code>.<br>
        The dimensions to which all images found will be resized.</li>
<li><strong>color_mode</strong>: one of "grayscale", "rgb", "rgba". Default: "rgb".<br>
        Whether the images will be converted to have 1 or 3 color channels.</li>
<li><strong>classes</strong>: optional list of classes (e.g. <code>['dogs', 'cats']</code>).<br>
        Default: None. If not provided, the list of classes will be
        automatically inferred from the <code>y_col</code>,
        which will map to the label indices, will be alphanumeric).
        The dictionary containing the mapping from class names to class
        indices can be obtained via the attribute <code>class_indices</code>.</li>
<li><strong>class_mode</strong>: one of "binary", "categorical", "input", "multi_output",<br>
        "raw", sparse" or None. Default: "categorical".
        Mode for yielding the targets:
        - <code>"binary"</code>: 1D numpy array of binary labels,
        - <code>"categorical"</code>: 2D numpy array of one-hot encoded labels.
            Supports multi-label output.
        - <code>"input"</code>: images identical to input images (mainly used to
            work with autoencoders),
        - <code>"multi_output"</code>: list with the values of the different columns,
        - <code>"raw"</code>: numpy array of values in <code>y_col</code> column(s),
        - <code>"sparse"</code>: 1D numpy array of integer labels,
        - <code>None</code>, no targets are returned (the generator will only yield
            batches of image data, which is useful to use in
            <code>model.predict_generator()</code>).</li>
<li><strong>shuffle</strong>: whether to shuffle the data (default: True)<br></li>
<li><strong>seed</strong>: optional random seed for shuffling and transformations.<br></li>
<li><strong>save_to_dir</strong>: Optional directory where to save the pictures<br>
        being yielded, in a viewable format. This is useful
        for visualizing the random transformations being
        applied, for debugging purposes.</li>
<li><strong>save_prefix</strong>: String prefix to use for saving sample<br>
        images (if <code>save_to_dir</code> is set).</li>
<li><strong>save_format</strong>: Format to use for saving sample images<br>
        (if <code>save_to_dir</code> is set).</li>
<li><strong>interpolation</strong>: Interpolation method used to resample the image if the<br>
        target size is different from that of the loaded image.
        Supported methods are <code>"nearest"</code>, <code>"bilinear"</code>, and <code>"bicubic"</code>.
        If PIL version 1.1.3 or newer is installed, <code>"lanczos"</code> is also
        supported. If PIL version 3.4.0 or newer is installed, <code>"box"</code> and
        <code>"hamming"</code> are also supported. By default, <code>"nearest"</code> is used.</li>
<li><strong>fit_sample_size</strong>: Int. Default is None / 1000.<br>
        Number of training images used in <code>datagen.fit</code>.
        Relevant only when <code>featurewise_center</code> or
        <code>featurewise_std_normalization</code> or <code>zca_whitening are set</code> are set
        to True.</li>
</ul>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="../feature-selectors/" class="btn btn-neutral float-right" title="feature_selectors">Next <span class="icon icon-circle-arrow-right"></span></a>
      
      
        <a href="../.." class="btn btn-neutral" title="Home"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
          <a href="https://github.com/goeckslab/Galaxy-ML/" class="fa fa-github" style="float: left; color: #fcfcfc"> GitHub</a>
      
      
        <span><a href="../.." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
        <span style="margin-left: 15px"><a href="../feature-selectors/" style="color: #fcfcfc">Next &raquo;</a></span>
      
    </span>
</div>
    <script>var base_url = '../..';</script>
    <script src="../../js/theme.js" defer></script>
      <script src="../../search/main.js" defer></script>

</body>
</html>
